src/Simulation/Simulators/QuantumProcessor/Return.cs

```diff
@@ -16,16 +16,18 @@ public class QuantumProcessorDispatcherReturn : Intrinsic.Return
 
             public override void Apply(Qubit q)
             {
-                long count = (sim.QubitManager.ToBeReleasedAfterReturn(q) ? 1 : 0);
-                sim.QuantumProcessor.OnReturnQubits(new QArray<Qubit>(q), count);
+                long allocatedBefore = sim.QubitManager.GetAllocatedQubitsCount();
                 sim.QubitManager.Return(q);
+                long allocatedAfter = sim.QubitManager.GetAllocatedQubitsCount();
+                sim.QuantumProcessor.OnReturnQubits(new QArray<Qubit>(q), allocatedBefore - allocatedAfter);
```

Comment:
A broader question: If we think there is value to these events in the general case, then I would move the events into SimulatorBase and move firing them into the qubit manager. If they're special-purpose events that we don't want to make generally available, then we can leave them here.

Reply:
There are already such events in the SimulatorBase. The QuantumProcessor has a method OnReturnQubits that it adds as handler to the corresponding OnReturnQubits event spawned by the simulator base.

Reply:
OK... I kind of hate chained event handlers, but that's *way* out of scope of this PR.

Reply:
This additional indirection should become obsolete as part of the larger refactoring, such that only the events in the SimulatorBase remain.